graveler/committed/diff.go

```diff
@@ -0,0 +1,207 @@
+	comp := d.compareDiffKeys()
+	switch {
+	case leftStartRange && rightStartRange && comp == 0:
+		return diffItCompareResultNeedStartRangeBoth
+	case leftStartRange && comp <= 0:
+		return diffItCompareResultNeedStartRangeLeft
+	case rightStartRange && comp >= 0:
+		return diffItCompareResultNeedStartRangeRight
+	case comp == 0 && bytes.Equal(d.leftVal.record.Identity, d.rightVal.record.Identity):
+		return diffItCompareResultSameIdentities
+	case comp == 0:
+		return diffItCompareResultSameKeys
+	case comp < 0:
+		return diffItCompareResultLeftBeforeRight
+	default:
+		return diffItCompareResultRightBeforeLeft
+	}
+}
+
+func (d *diffIterator) Next() bool {
+	if d.state == diffItClosed {
+		return false
+	}
+	if d.state == diffItBeforeInit {
+		d.state = diffItInitialized
+		d.leftVal.record, d.leftVal.rng, d.err = d.diffIteratorNextValue(d.left)
+		d.rightVal.record, d.rightVal.rng, d.err = d.diffIteratorNextValue(d.right)
+	}
+	for {
+		if d.err != nil {
+			return false
+		}
+		compareResult := d.compareDiffIterators()
+		switch compareResult {
+		case diffItCompareResultDone:
+			d.currentVal = nil
+			return false
+		case diffItCompareResultSameRanges:
+			d.leftVal.rng, d.err = d.diffIteratorNextRange(d.left)
+			d.rightVal.rng, d.err = d.diffIteratorNextRange(d.right)
+		case diffItCompareResultSameKeys:
+			// same keys on different ranges
+			d.currentVal = d.rightVal.asDiff(graveler.DiffTypeChanged)
+			d.leftVal.record, d.leftVal.rng, d.err = d.diffIteratorNextValue(d.left)
+			d.rightVal.record, d.rightVal.rng, d.err = d.diffIteratorNextValue(d.right)
+			return true
+		case diffItCompareResultSameIdentities, diffItCompareResultNeedStartRangeBoth:
+			d.leftVal.record, d.leftVal.rng, d.err = d.diffIteratorNextValue(d.left)
+			d.rightVal.record, d.rightVal.rng, d.err = d.diffIteratorNextValue(d.right)
+		case diffItCompareResultNeedStartRangeLeft:
+			d.leftVal.record, d.leftVal.rng, d.err = d.diffIteratorNextValue(d.left)
+		case diffItCompareResultNeedStartRangeRight:
+			d.rightVal.record, d.rightVal.rng, d.err = d.diffIteratorNextValue(d.right)
+		case diffItCompareResultLeftBeforeRight:
+			// nothing on right, or left before right
+			d.currentVal = d.leftVal.asDiff(graveler.DiffTypeRemoved)
+			d.leftVal.record, d.leftVal.rng, d.err = d.diffIteratorNextValue(d.left)
+			return true
+		case diffItCompareResultRightBeforeLeft:
+			// nothing on left, or right before left
+			d.currentVal = d.rightVal.asDiff(graveler.DiffTypeAdded)
+			d.rightVal.record, d.rightVal.rng, d.err = d.diffIteratorNextValue(d.right)
+			return true
+		}
+	}
+}
+
+func (d *diffIterator) SeekGE(id graveler.Key) {
+	d.left.SeekGE(id)
+	d.right.SeekGE(id)
+	d.currentVal = nil
+	d.leftVal = &IteratorValue{}
+	d.rightVal = &IteratorValue{}
+	d.err = nil
+	d.state = diffItBeforeInit
+}
+
+func (d *diffIterator) Value() *graveler.Diff {
+	return d.currentVal
+}
+
+func (d *diffIterator) Err() error {
+	return d.err
+}
+
+func (d *diffIterator) Close() {
```

Comment:
can check if the current state is already closed and return if you like the graceful behavior